processing/src/main/java/org/apache/druid/query/QueryProcessingPool.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.apache.druid.guice.annotations.ExtensionPoint;
+
+/**
+ * This class implements the logic of how units of query execution run concurrently. It is used in {@link QueryRunnerFactory#mergeRunners(QueryProcessingPool, Iterable)}.
+ * In a most straightforward implementation, each unit will be submitted to an {@link PrioritizedExecutorService}. Extensions,
+ * however, can implement their own logic for picking which unit to pick first for execution.
+ * <p>
+ * This interface is convertible to a regular {@link java.util.concurrent.ExecutorService} as well. It has a separate
+ * method to submit query execution tasks so that implementations can differentiate those tasks from any regular async
+ * tasks. One example is {@link org.apache.druid.query.groupby.strategy.GroupByStrategyV2#mergeRunners(QueryProcessingPool, Iterable)}
+ * where different kind of tasks are submitted to same processing pool.
+ * <p>
+ * Query execution task also includes a reference to {@link QueryRunner} so that any state required to decide the priority
+ * of a unit can be carried forward with the corresponding {@link QueryRunner}.
+ */
+@ExtensionPoint
+public interface QueryProcessingPool
+{
+  /**
+   * Submits the query execution unit task for asynchronous execution.
+   *
+   * @param task - Task to be submitted.
+   * @param <T>  - Task result type
+   * @param <V>  - Query runner sequence type
+   * @return - Future object for tracking the task completion.
+   */
+  <T, V> ListenableFuture<T> submit(PrioritizedQueryRunnerCallable<T, V> task);
+
+  /**
+   * @return - Returns this pool as an executor service that can be used for other asynchronous operations.
+   */
+  ListeningExecutorService asExecutorService();
```

Comment:
I think this method leaks the internals of the interface. For instance, some caller may call this method and then pass the executor in other parts of code unintentionally. Also, if someone has to implement a `QueryProcessingPool` which is composite and contains multiple pools inside it, it would become hard to implement this interface.
Would it be better to rather have the original `ExecutorService` as is, and then inject that executor service to the `DefaultProcessingPool`? The `ExecutorService` interface is richer and common. `QueryProcessingPool` can be used in cases where a `PrioritizedQueryRunnerCallable` submission is needed.

Reply:
something like this?
```
@ExtensionPoint
public interface QueryProcessingPool extends ListeningExecutorService
{
  /**
   * Submits the query execution unit task for asynchronous execution.
   *
   * @param task - Task to be submitted.
   * @param <T>  - Task result type
   * @param <V>  - Query runner sequence type
   * @return - Future object for tracking the task completion.
   */
  <T, V> ListenableFuture<T> submitQueryExecution(PrioritizedQueryRunnerCallable<T, V> task);
}
```

Reply:
Yes, I was thinking like the snippet you mentioned above. (probably the interface can be called something like `QueryRunnerProcessingPool` since it only allows submit for `QueryRunners` currently - the `submitQueryExecution` can also become `submit`).
If this interface looks ok, then the DefaultProcessingPool's constructor signature can be : `DefaultProcessingPool(@Processing ExecutorService)`

Reply:
I assume you want to make `QueryProcessingPool` compatible with `ExecutorService` because of `ConcurrentGrouper`?

Reply:
Just clarifying my stance here, @rohangarg's idea sounds good to me, but I don't have strong preference here.

Reply:
I think making the processing a pool a kind of executor service makes sense, let's go with that if you agree

Reply:
I do want to keep the method separate though. Either that or `PrioritizedQueryRunnerCallable<T, V> task` should not extend `Callable`. My reasoning is that then Implementations don't have to do `instance of` checks for differentiating between query execution tasks and other async tasks. 

Yes
> I assume you want to make QueryProcessingPool compatible with ExecutorService because of ConcurrentGrouper?

Reply:
@rohangarg @jon-wei @jihoonson What do you think?

Reply:
LGTM
<nit> maybe the naming could be more generic since this would be used in both ingestion and querying layer both.

Reply:
On real-time nodes too, the pool is used in query execution.

Reply:
SGTM